src/pages/chartPage.js

```diff
@@ -1,12 +1,276 @@
+            moment(today).add(2, "days"),
+          );
+          setRecords(newRecords);
+          setLegend("-Day (1 week ago)");
+        }}/>
+      <Button
+      title="Fetch data a month ago"
+      icon="cloud-download-outline"
+      onPress={async () => {
+        const newRecords = await RecordStore.getRange(
+          moment(today).subtract(30, "days"),
+          moment(today).add(2, "days"),
+        );
+        setRecords(newRecords);
+        setLegend("-Day (1 month ago)");
+      }}/>
+      <Button
+      title="Fetch data 3 months ago"
+      icon="cloud-download-outline"
+      onPress={async () => {
+        const newRecords = await RecordStore.getRange(
+          moment(today).subtract(90, "days"),
+          moment(today).add(2, "days"),
+        );
+        setRecords(newRecords);
+        setLegend("-Day (3 months ago)");
+      }}/>
+      <ChartCanvus
+        records = {records}
+        legend = {"Weight"+legend}
+        yKey = {"weight"}
+        interval = {10}
+        ySuffix = {"kh"}
+      />
+      <ChartCanvus
+        records = {records}
+        legend = {"Fat"+legend}
+        yKey = {"fat"}
+        interval = {1}
+        ySuffix = {"%"}
+      />
+    </View>
+  )
+  
+         
+}
+
+const ChartCanvus = (props) => {
+  const { records, legend = "Weight-Month", yKey='weight', interval=50, ySuffix="kg"} = props;
+
+  const [focusLocation, setFocusLocation] = useState(null);
+  // Use tmpDotArray only when rendering in the fist time.
+  const tmpDotXArray = []; 
+  const [dotXArray, setDotXArray] = useState(null);
+  
+  // Compute the closest dot when focusLocation changing.
+  const [closestDotIdx, setClosestDotIdx] = useState(null);
+  useEffect(() => {
+    setClosestDotIdx(findClosestDot(dotXArray, focusLocation));
+  }, [focusLocation]);
+  
+  // When closest dot changing, modify the xHidden array to show less dot and avoid wasting too much effort.
+  const [focusRecords, setFocusRecords] = useState({date: "", weight: "", fat: ""});
+  const [xHidden, setXHidden] = useState([]);
+  useEffect(() => {
+    if (!_.isNull(closestDotIdx)) {
+      setXHidden(_.range(0, closestDotIdx).concat(_.range(closestDotIdx+1, dotXArray.length)));
+      setFocusRecords({
+        date: moment(props.records[closestDotIdx]["date"]).format("YYYY-MM-DD"),
+        weight: props.records[closestDotIdx]["weight"].toFixed(2),
+        fat: props.records[closestDotIdx]["fat"].toFixed(2),
+      });
+    }
+  }, [closestDotIdx]);
+
+  const resetAllState = () => {
+    setFocusLocation(null); 
+    setClosestDotIdx(null);
+    setXHidden([]);
+    setFocusRecords({date: "", weight: "", fat: ""});
+  }
+  const saveDotXArrayInFirstTime = (tmpArray) => {
+    if(_.isNull(dotXArray)){
+      setDotXArray(tmpArray);
+    }
+  }
+
+  //Computer yMax and yMin
+  const yMax = (props.records.length>0)?Math.ceil(_.max(props.records.map((d)=>d[props.yKey]))/props.interval+1)*props.interval : 100;
```

Comment:
please don't do this. It's so hard to understand.

Reply:
Please tell me how to write more understandable code on these.